drivers/dma/dma_sedi.c

```diff
@@ -0,0 +1,533 @@
+#include <errno.h>
+#include <stdio.h>
+#include <zephyr/kernel.h>
+#include <zephyr/pm/device.h>
+#include <string.h>
+#include <zephyr/init.h>
+#include <zephyr/drivers/dma.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/cache.h>
+#include <soc.h>
+
+#include "sedi_driver_dma.h"
+#include "sedi_driver_core.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(sedi_dma, CONFIG_DMA_LOG_LEVEL);
+
+extern void dma_isr(sedi_dma_t dma_device);
+
+struct dma_sedi_config_info {
+	sedi_dma_t peripheral_id; /* Controller instance. */
+	uint8_t chn_num;
+};
+
+struct dma_sedi_driver_data {
+	struct dma_config dma_configs[DMA_CHANNEL_NUM];
+	uint8_t power_status;
+};
+
+#define DEV_DATA(dev) ((struct dma_sedi_driver_data *const)(dev)->data)
+#define DEV_CFG(dev) \
+	((const struct dma_sedi_config_info *const)(dev)->config)
+
+/*
+ * this function will be called when dma transferring is completed
+ * or error happened
+ */
+static void dma_handler(sedi_dma_t dma_device, int channel, int event_id,
+			void *args)
+{
+	ARG_UNUSED(args);
+	const struct device *dev = (const struct device *)args;
+	struct dma_sedi_driver_data *const data = DEV_DATA(dev);
+	struct dma_config *config = &(data->dma_configs[channel]);
+
+	/* run user-defined callback */
+	if (config->dma_callback) {
+		if ((event_id == SEDI_DMA_EVENT_TRANSFER_DONE) &&
+		    (config->complete_callback_en)) {
+			config->dma_callback(dev, config->user_data,
+					channel, 0);
+		} else if (config->error_callback_en) {
+			config->dma_callback(dev, config->user_data,
+					channel, event_id);
+		}
+	}
+}
+
+/* map width to certain macros*/
+static int width_index(uint32_t num_bytes, uint32_t *index)
+{
+	switch (num_bytes) {
+	case 1:
+		*index = DMA_TRANS_WIDTH_8;
+		break;
+	case 2:
+		*index = DMA_TRANS_WIDTH_16;
+		break;
+	case 4:
+		*index = DMA_TRANS_WIDTH_32;
+		break;
+	case 8:
+		*index = DMA_TRANS_WIDTH_64;
+		break;
+	case 16:
+		*index = DMA_TRANS_WIDTH_128;
+		break;
+	case 32:
+		*index = DMA_TRANS_WIDTH_256;
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+/* map burst size to certain macros*/
+static int burst_index(uint32_t num_units, uint32_t *index)
+{
+	switch (num_units) {
+	case 1:
+		*index = DMA_BURST_TRANS_LENGTH_1;
+		break;
+	case 4:
+		*index = DMA_BURST_TRANS_LENGTH_4;
+		break;
+	case 8:
+		*index = DMA_BURST_TRANS_LENGTH_8;
+		break;
+	case 16:
+		*index = DMA_BURST_TRANS_LENGTH_16;
+		break;
+	case 32:
+		*index = DMA_BURST_TRANS_LENGTH_32;
+		break;
+	case 64:
+		*index = DMA_BURST_TRANS_LENGTH_64;
+		break;
+	case 128:
+		*index = DMA_BURST_TRANS_LENGTH_128;
+		break;
+	case 256:
+		*index = DMA_BURST_TRANS_LENGTH_256;
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static struct dma_block_config *
+alloc_and_cpy_blk_nodes(struct dma_block_config *src_blk)
+{
+	int i = 0;
+	struct dma_block_config *blk_p, *blk_p_h = NULL;
+
+	while (src_blk) {
+		blk_p = (struct dma_block_config *) k_malloc(sizeof(struct dma_block_config));
```

Comment:
k_malloc generally frowned upon in this case. Since its a fixed size item you can...

* Use a object pool dedicated to the DMA driver
* Statically allocate a fixed number of blocks per channel using them as needed

Either is acceptable, k_malloc is generally not. See the nxp edma or designware dma drivers for some samples of how this can be done

Reply:
No link list support here, will rewrite this code here.